### PR TITLE
ignore [no test files]

### PR DESCRIPTION
**Pull request by mostafasolati**
This change removes the packages that have no test files by default.
We still can see those package in verbose mode (-v) but seeing these lines in every test is annoying so I removed them by default.

### CODE REVIEW

**Comment by subtlepseudonym** [comment]
This seems like behavior that should be controlled by adding a flag (like `--omit-ignored`) rather than changing the default behavior.

**Comment by rakyll** [comment]
I agree with @subtlepseudonym that we should instead have a separate flag and shouldn't change the default behavior of "go test".

**Comment by mostafasolati** [comment]
> This seems like behavior that should be controlled by adding a flag (like `--omit-ignored`) rather than changing the default behavior.

Okay I added a flag called -ignore and its default value is false

File: main.go [hunk 1]
```diff
@@ -23,15 +24,16 @@ import (
 var (
 	success = color.New(color.FgGreen)
 	fail    = color.New(color.FgHiRed)
-	ignore = true
+	ignore = flag.Bool("ignore",false,"ignore packages with no test files")
```
Comment (subtlepseudonym) [inline main.go:26]: I don't believe `-ignore` is a descriptive enough flag name.

File: main.go [hunk 2]
```diff
@@ -23,15 +24,16 @@ import (
...
 )
 
 const paletteEnv = "GOTEST_PALETTE"
 
 func main() {
+	flag.Parse()
 	setPalette()
 	enableOnCI()
-	os.Exit(gotest(os.Args[1:]))
+	os.Exit(gotest(os.Args[2:]))
```
Comment (subtlepseudonym) [inline main.go:39]:
```suggestion
	os.Exit(gotest(os.Args[1:]))
```

We don't know that the flag is present or that it's one of the first two arguments. Additionally, `go test` deals with unknown flags gracefully, so there's no reason to avoid forwarding it.

File: main.go (hunk 1 above)
Comment (mostafasolati) [reply main.go:26]: What is your suggestion?

Comment (mostafasolati) [reply main.go:26]: I changed the name to -skipnotest

File: main.go [hunk 3]
```diff
@@ -22,17 +23,26 @@ import (
...
 	enableOnCI()
-	os.Exit(gotest(os.Args[1:]))
+
+	var startingIndex = 1
+	if strings.HasPrefix(os.Args[1], "-skipnotest") {
+		startingIndex++
+	}
+
+	os.Exit(gotest(os.Args[startingIndex:]))
+
```
Comment (subtlepseudonym) [inline main.go:45]:
```suggestion

	os.Exit(gotest(os.Args[1:]))
```

I don't believe this additional logic adds anything to `gotest`. So long as we don't reuse any flag names that `go test` uses, it's fine passing them along.
For example, running `go test --skipnotest ./...` doesn't error and simply ignores the unknown flag.

File: main.go [hunk 4]
```diff
@@ -103,9 +113,11 @@ func parse(line string) {
 	case strings.HasPrefix(trimmed, "=== RUN"):
 		fallthrough
 	case strings.HasPrefix(trimmed, "?"):
+		if *ignore {
+			return
+		}
 		color.Unset()
 
-	// passed
```
Comment (subtlepseudonym) [inline main.go:108]: Why was this removed?

**Comment by subtlepseudonym** [comment]
@mostafasolati , if you're open to it, I have a few suggestions on the use of git that will make your work clearer.

File: main.go (hunk 4 above)
Comment (mostafasolati) [reply main.go:108]: It was a mistake.

File: main.go (hunk 3 above)
Comment (mostafasolati) [reply main.go]: I changed HasPrefix to Contains is that good?

File: main.go [hunk 5]
```diff
@@ -22,17 +23,26 @@ import (
...
 	setPalette()
 	enableOnCI()
-	os.Exit(gotest(os.Args[1:]))
+
+	var startingIndex = 1
+	if strings.Contains(os.Args[1], "-skipnotest") {
+		startingIndex++
+	}
+
+	os.Exit(gotest(os.Args[startingIndex:]))
```
Comment (rakyll) [inline main.go:44]: You can use flag.Args instead, see https://golang.org/pkg/flag/#Args

It returns the non-flag arguments.

Comment (mostafasolati) [reply main.go:44]: Wonderful!

Comment (subtlepseudonym) [reply main.go]: I think using `flag.Args` will prevent flags that we intend to pass to `go test` from being forwarded.

**Comment by subtlepseudonym** [comment]
@mostafasolati you can find the git docs [here](https://git-scm.com/doc)

I've tried to summarize a few helpful points below --

- `git add some_file.go` makes isolating your changes a lot easier and won't add things like `.idea/`
    - This will also help if you want to add `.idea/` to your `.gitignore` file, but don't want to commit the `.gitignore`
- `git add -p .` will give you even greater granularity, splitting things up into "hunks" and using an interactive mode to let you decide what to do with each hunk individually
- To unstage things you've added, use `git reset`
- Try separating commits into small, related changes
- Commit messages should indicate what that specific commit does to the codebase
    - `use flag.Args instead of os.Args` is a great commit message because it summarizes the commit succinctly
    - `some changes` isn't a very good message because I now have to look at the changes to know what it does
- If you commit something you didn't intend or otherwise want to update the commit, you can use `git commit --amend`, though this is only relevant if you haven't pushed your changes yet

Let me know if anything is unclear or unhelpful.

File: main.go (hunk 5 above)
Comment (mostafasolati) [reply main.go]: @subtlepseudonym 
No, It doesn't I tested it.
for example:
`gotest -skipnotest ./... -cover
`

**Comment by mostafasolati** [comment]
@subtlepseudonym 
What does the "Some checks haven't completed yet" message mean?

**Comment by subtlepseudonym** [comment]
@mostafasolati I suspect you're getting that error from attempting to merge this PR. 

Specifically, the error means that the `ci/circleci: build` check below has not completed successfully and github won't merge the pull request until it has. More generally, you can't merge this PR for two reasons: it hasn't been approved and you don't have write permissions.

**Comment by mostafasolati** [comment]
> @mostafasolati I suspect you're getting that error from attempting to merge this PR.
> 
> Specifically, the error means that the `ci/circleci: build` check below has not completed successfully and github won't merge the pull request until it has. More generally, you can't merge this PR for two reasons: it hasn't been approved and you don't have write permissions.

Wierd, how can I attempt to merge without permission?

File: main.go [hunk 6]
```diff
@@ -22,17 +23,21 @@ import (
...
+	pass   = color.FgGreen
+	skip   = color.FgYellow
+	fail   = color.FgHiRed
+	ignore = flag.Bool("skipnotest", false, "skip packages with no test files")
 )
 
 const paletteEnv = "GOTEST_PALETTE"
 
 func main() {
+	flag.Parse()
```
Comment (subtlepseudonym) [inline main.go:35]: Using `flag.Parse` causes unexpected behavior in `go test`. For example, obtaining verbose output from `go test` requires the `-v` flag to be used after non-flag arguments. Specifically,
`gotest ./... -v` works, but `gotest -v ./...` doesn't.

We should use a `flag.FlagSet` with `flag.ContinueOnError` so that we can avoid exiting when we attempt to parse flags intended for `go test`.

Comment (subtlepseudonym) [reply main.go:35]: Unfortunately, it doesn't look like the flag package provides a typed error for unknown flags, so we're stuck with string parsing, which I'm not a huge fan of for error identification.

https://golang.org/src/flag/flag.go?s=31609:31672#L935

**Comment by subtlepseudonym** [comment]
> Sounds wierd, how can I attempt to merge without permission?

You can't, which is by design. This is @rakyll 's repo and, to my knowledge, they are the only one with write permission. They have the final say on whether your changes are accepted and merged.

File: main.go (hunk 6 above)
Comment (subtlepseudonym) [reply main.go:35]: Even with `ContinueOnError`, flag parsing is halted on the first unknown flag.

File: main.go [hunk 7]
```diff
@@ -22,17 +23,43 @@ import (
...
 )
 
 const paletteEnv = "GOTEST_PALETTE"
 
 func main() {
 	setPalette()
 	enableOnCI()
-	os.Exit(gotest(os.Args[1:]))
+
+	flagSet := flag.NewFlagSet("gotestFlags",flag.ContinueOnError)
```
Comment (rakyll) [inline main.go:43]: Please gofmt the code.

File: main.go [hunk 8]
```diff
@@ -22,17 +23,43 @@ import (
...
+	ignore=flagSet.Bool("skipnotest", false, "skip packages with no test files")
```
Comment (rakyll) [inline main.go:44]: Move parseFlags here.

var parseFlags []string

I'd also name it to something like gotestFlags

File: main.go [hunk 9]
```diff
@@ -22,17 +23,43 @@ import (
...
+
+	// separate program specific flags from go test flags
+	for _,arg:=range os.Args[1:] {
+		for _,flg := range gotestFlags {
```
Comment (rakyll) [inline main.go:48]: Instead of this, you can use https://golang.org/pkg/flag/#FlagSet.Lookup

**Comment by rakyll** [comment]
@mostafasolati please ignore the CircleCI, I can merge the change without it passing. I have been trying to set it up but the builds seem to be stuck. Please don't worry about it. If you can address the comments, I can merge :)

File: main.go (hunk 7 above)
Comment (mostafasolati) [reply main.go:43]: What do you mean? It's been formatted.

File: main.go (hunk 8 above)
Comment (mostafasolati) [reply main.go:44]: there is a gotestFlags variable already!

**Comment by rakyll** [comment]
The filtering of arguments and flags were causing a few bugs and is harder for us to maintain due to the lack of comprehensive testing. Instead, I switched to an environment variable to make things easier to maintain. https://github.com/rakyll/gotest/issues/30
